### DISCUSSION TITLE
hard-coded vs. dynamic BT target device name

### DISCUSSION THREAD

**Discussion by korkbaum** [pschatzmann/ESP32-A2DP#486] [Q&A] [answered]
Hello Phil,
with the latest version of your impressing libraries it was really straightforward to get an ESP32-A1S (AI Audiokit 2.2) running with Audio Line-In input and send the audio to a BT headset in very good quality. 
The only change that was needed was the name of the BT headset in the code and that is my question: 
How can the hardcoded name be avoided and replaced with a choice of BT names coming from the BT discovery? 
I can see my different BT devices in Core Debug messages, but filter_inquiry_scan_result() will only jump on the one with the hardcoded name.
I'm familiar with running external displays with an esp32 and adding some buttons for making a choice also shouldn't be a problem at all. The part I'm struggling with is how to start a BT scan, getting the resulting BT device list in the esp32 setup/loop, then making a choice and run a2dp_source.start with it. 
Can you please give me some direction on what the rquired changes would be? 
Many Thanks & Best Regards
Klaus

**Comment by pschatzmann** [comment, marked as answer]
I am afraid I don't quite understand what scenario are you trying to implement: a source or a sink ?
If you talk about a source, why don't you just provide the list of all devices ? Or you would register a callback method via set_ssid_callback which provides you with the discovered names.

**Comment by korkbaum** [comment]
Apologies that the description was so vague.

The signal chain is: Audio line out from an iPod -> Line in of AI Audiokit 2.2 -> Internal redirection to BT A2DP source -> send audio to BT headset/speaker.

That all works. What bothers me is that the name of the receiving BT headset/speaker is hard-coded and I would wish to get a list of compatible BT devices in range.

So what would need to be done is 
a - start a device discovery from the main code
b - send the list of found A2DP BT devices (potential sinks) back
c - let the user make a choice from that list somehow
d - start a2dp_source with that device name

So your guess was right "If you talk about a source, why don't you just provide the list of all devices? Or you would register a callback method via set_ssid_callback which provides you with the discovered names."
Many thanks for the hint. I will try my best to get this done although it might go beyond my capabilities. 

Thanks & BR, Klaus

**Comment by pschatzmann** [comment]
If you are a beginner, I suggest that you try to use the easy way by just providing a vector of all your devices. You can google how to initialize a std::vector!
The collection and building of a list is quite tricky because the callback just provides you individual found devices one by one and you would need to ignore the ones which are reported subsequently again.

Further details can be found in the class documentation.

**Comment by korkbaum** [comment]
Thanks, Phil, I will give it a try.
